libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs

```diff
@@ -90,15 +90,15 @@ public object EvaluateTemplate(string inputTemplateName, object scope)
                 throw new Exception(TemplateErrors.TemplateNotExist(templateName));
             }
 
-            if (evaluationTargetStack.Any(e => e.TemplateName == templateName))
-            {
-                throw new Exception($"{TemplateErrors.LoopDetected} {string.Join(" => ", evaluationTargetStack.Reverse().Select(e => e.TemplateName))} => {templateName}");
-            }
-
             var templateTarget = new EvaluationTarget(templateName, scope);
 
             var currentEvaluateId = templateTarget.GetId();
 
+            if (evaluationTargetStack.Any(e => e.GetId() == currentEvaluateId))
```

Comment:
>GetId [](start = 49, length = 5)

This solution looks to work, but I'm not fond of the underlying implementation of GetId for CustomMemory.  It should probably use Version(), instead it calls ToString() on every call--this is not part of IMemory and kind of only works because we happen to use SimpleObjectMemory which has a ToString which serializes the object. A better implementation would do ToString() inside of Version() and use Version() in GetId() for CustomMemory.

Reply:
Let me explain the GetId of EvaluationTarget,
1. All scope in Evaluation is CustomizedMemory, because LG makes the conversion at the beginning of the `EvaluateTemplate` method.
1. The implements of GetId is from `version`, the code is :
```
public string GetId()
        {
            var memory = (CustomizedMemory)Scope;
            var globalMemoryId = memory.GlobalMemory?.Version() ?? string.Empty;
            var localMemoryId = memory.LocalMemory?.ToString() ?? string.Empty;
            return TemplateName + globalMemoryId + localMemoryId;
        }
```

Use Tostring as the `SimpleObjectMemory` version implement, maybe is a doable solution.

Reply:
If i understand correctly, @chrimc62  is suggesting the implementation of GetId should only depends on the abstraction of `Version()` which is part of the interface, i agree, this is neat, and don't have loophole in the abstraction. 

We should keep the toString trick as an implementation of version() inside SimpleObjectMemory, and also if better cache it, so you don't have to do it everytime. @Danieladu

Reply:
LGTM now